src/graph.jl

```diff
@@ -0,0 +1,349 @@
+
+    v_df = DataFrame(include(vertices_file))
+    
+    # Case 1: the graph is currently empty; objective is to create it from the input vertex and edge files.
+    if isa(input_graph, Nothing)
+        
+        n_vertices = size(v_df,1)
+        G = MetaDiGraph(n_vertices)
+        counter = 1::Int64
+
+        set_indexing_prop!(G, :v_hash)
+        
+        for v in eachrow(v_df)
+        
+            v_hash = hash(string(v.v_name, v.v_type)) 
+            
+            v_attrs = Dict(
+                :v_name=>v.v_name,
+                :v_type=>v.v_type,
+                :used=>true)
+            
+            set_indexing_prop!(G, counter, :v_hash, "$v_hash")
+            set_props!(G, counter, v_attrs)
+            @info "Inserting new vertex into graph: $v"
+            counter += 1
+        end
+    
+    # Case 2: the graph already exists (as a graph object); objective is to add new vertices/edges, and increment weights when existing edges are encountered.
+    else
+        
+        g_input = copy(input_graph)
+        n_vertices_input = nv(g_input)
+        n_vertices = size(v_df,1) + n_vertices_input
+        
+        counter = nv(g_input) + 1
+
+        G = MetaDiGraph(n_vertices)
+        
+        # insert vertices (with properties) from the input graph
+        for i in 1:nv(g_input)
+            
+            v_props = props(g_input, i)
+            
+            set_indexing_prop!(G, i, :v_hash, v_props[:v_hash])
+            
+            set_props!(G, i, 
+                Dict(:v_name=>v_props[:v_name], 
+                    :v_type=>v_props[:v_type],
+                    :used=>true))
+        end
+        
+        @info "$n_vertices_input from input graph have been inserted."
+        
+        for v in eachrow(v_df)
+        
+            v_hash = hash(string(v.v_name, v.v_type)) 
+
+            try 
+                G["$v_hash", :v_hash]
+                
+                @info "A vertex with this hash value already exists"
+                
+                # Set the "used" attribute to false, so we can remove this vertex before returning G
+                 v_attrs = Dict(
+                        :v_name=>v.v_name,
+                        :v_type=>v.v_type,
+                        :used=>false)
+            
+                    set_indexing_prop!(G, counter, :v_hash, string("$v_hash", "_copy"))
+                    set_props!(G, counter, v_attrs)
+                    
+            catch error
+                if isa(error, KeyError) || isa(error, ExceptionError)
+                
+                v_attrs = Dict(
+                        :v_name=>v.v_name,
+                        :v_type=>v.v_type,
+                        :used=>true)
+            
+                    set_indexing_prop!(G, counter, :v_hash, "$v_hash")
+                    set_props!(G, counter, v_attrs)
+                    
+                    @info "Inserting new vertex into graph: $v."
```

Comment:
You can make structured logs by passing kwargs to `@info`

`@info("Inserting new vertex into graph", vertex=v)`

These structured logs can be written out in JSON to make parsing easier.

Reply:
will work on re-structuring / piping to JSON in next iteration.